README.md

```diff
@@ -39,46 +40,62 @@ Concretely, it's a *typed* content address: a tuple of `(content-type, content-a
 - `<multicodec-cidv1>` is a [multicodec](https://github.com/multiformats/multicodec) representing the version of CID, here for upgradability purposes.
 - `<multicodec-content-type>` is a [multicodec](https://github.com/multiformats/multicodec) code representing the content type or format of the data being addressed.
-- `<multihash-content-address>` is a [multihash](https://github.com/multiformats/multihash) value, representing the cryptographic hash of the content being addressed. Multihash enables CIDs to use many different cryptographic hash function, for upgradability and protocol agility purposes.
+- `<multihash-content-address>` is a [multihash](https://github.com/multiformats/multihash) value, which uses a registry of hash function abbreviations to prefix a cryptographic hash of the content being addressed, thus making it self-describing.
 
-That's it!
+## Variant - Stringified Form
 
-## Design Considerations
+Since CIDs have many applications outside of binary-only contexts, a given CID may need to be base-encoded multiple ways for different consumers or for different transports. 
+In such applications, CIDs are often expressed as a Unicode *string* rather than a bytestring, which adds a single code-point prefix. 
+In these contexts, then, the full string form is:
 
-CIDs design takes into account many difficult tradeoffs encountered while building [IPFS](https://ipfs.io). These are mostly coming from the multiformats project.
+```text
+<cidv1> ::= <multibase-codec><multibase-encoding(<CIDv1-multicodec><content-multicodec><multihash>)>
+```
 
-- Compactness: CIDs are binary in nature to ensure these are as compact as possible, as they're meant to be part of longer path identifiers or URIs.
-- Transport friendliness (or "copy-pastability"): CIDs are encoded with multibase to allow choosing the best base for transporting. For example, CIDs can be encoded into base58btc to yield shorter and easily-copy-pastable hashes.
-- Versatility: CIDs are meant to be able to represent values of any format with any cryptographic hash.
-- Avoid Lock-in: CIDs prevent lock-in to old, potentially-outdated decisions.
-- Upgradability: CIDs encode a version to ensure the CID format itself can evolve.
+Where
+
+- `<multibase-codec>` is a [multibase](https://github.com/multiformats/multibase) prefix  (1 Unicode codepoint in length) that renders the base-encoded unicode string following it self-describing for simpler conversion back to binary.
+ 
+[internal CID section](#variant-representation---internal-cid) below for context.
+
+## Variant - Legacy Form
+
+In some binary-only contexts, such as the binary CIDs used as links in the binary data structure [DAG-CBOR](https://ipld.io/specs/codecs/dag-cbor/spec/#links)), CIDs are encoded in binary, but with the one-byte "NULL" prefix (0x00) used like an "escape" character to signal to multibase consumers that what follows is NOT to be base-encoded.
```

Comment:
I believe dag-cbor is the _only_ place where this is done, and it's handled and stripped out at the codec layer before going anywhere else (i.e. the only code that has to consider this is the dag-cbor codecs themselves, exactly one place per language/implementation). I wonder if it's worth pointing this out and codifying it as an option that people may reach for? It's an interesting quirk of history, but not really something we'd want people reaching for when designing new systems.

Reply:
good counterpoint.  i think the big-picture question i'm trying to handle here is "how do the interlocking pieces fit together to parse a binary stream that isn't protobuf and isn't CBOR?", so I'm trying to build a pyramid (from lower levels to higher levers) where `unsigned varints` and multicodecs are the base and CIDs are the apex; this is why the specs for seemingly simple things like multibase and multihash are getting weirdly "defensive" and explaining themselves relative to what they're not :D

but @vmx 's point here was that the `identity` prefix is more like a non-multibase prefix, and the "external" binary of CBOR-land has a "don't read me as multibase" envelope, every other external transport/context has a "this is my multibase" envelope, and the "internal" binary contexts of IPFS `unsigned varint`-land do not need either... which was a convincing explanation to me of many quirks I was having trouble conceptualizing (as per the above paragraph!)

Reply:
@bumblefudge I'm sorry to confuse you. I'm on the same page as @rvagg here, he's better at describing it. I really think the Multibase should *not* be part of the CID at all. You would "Multibase encode CIDs", but it's not inherently part of the CID.

In DAG-CBOR, we have this `identity` prefix due to historic reasons (when Multibase was considered part of the CID). But as @rvagg points out, I think it should only be described there and not in the CID spec.

Reply:
I would just remove Variants "legacy form" and "CBOR" compatibility from this spec. I treat the NUL prefix just a CBOR tag 42 specific thing, which is there due to historic reasons (which don't really matter) and won't be changed as we don't want to make a breaking change.

So from a CBOR perspective it's just a CID with a NUL prefix, it doesn't matter that it used to be due to CID being defined differently in the past.

Reply:
Ah, OK, so the extra NUL isn't required, and a modern non-multibased, non-prefixed CID would get parsed the same way by a standard CBOR parser?

Reply:
> Ah, OK, so the extra NUL isn't required, and a modern non-multibased, non-prefixed CID would get parsed the same way by a standard CBOR parser?

No. CID has nothing to do with CBOR. Perhaps that's the confusion here. Yes, you can encode a CID in CBOR, but you can also encode a CID in MessagePack. The fact that CIDs are encoded in CBOR in a certain way should not be part of the CID spec.

Reply:
> I would also add that the NUL-prefix is still going to be a reserved and explained value in the Multibase spec at IETF, so I'm still struggling to explain its exact syntax and usage. Is this legacy usage from the days of mandatory multibase the only documented usage or are there others I can point to, whether as examples or by way of explanation?

The NUL-prefix isn't specific to CID, I think it's again out of scope for the CID spec. It only matters for the Multibase spec. It's used in case someone wants to have a self-describing binary encoding. The only usage I know is in CBOR CIDs.

Reply:
Yep, what @vmx said.

We consider this a quirk of history that's been locked in place and won't be changed, but that's only for the dag-cbor codec and doesn't apply anywhere else. My fear is that raising it here lifts it out of dag-cbor and makes it a more generic concern, which it isn't. It should just be something in the dag-cbor spec. The entire section here https://github.com/ipld/ipld/blob/master/specs/codecs/dag-cbor/spec.md#links really should be that last sentence. Essentially: "Hey, this was a historical mistake but it is what it is, CIDs in dag-cbor have an extra `0x00` in front of them, an encoder needs to put one there and a decoder needs to expect one there."

I'm not sure the idea of an "identity" encoding of binary CIDs really helps us at all, at least I haven't seen a case (I think it was originally thought it would be useful when drafting dag-cbor, but we quickly found it was pointless and just wastes a byte).

> I would also add that the NUL-prefix is still going to be a reserved and explained value in the Multibase spec at IETF

Is this necessary beyond just saying `0x00` is a reserved value as a Multibase prefix and shouldn't be used?

Reply:
ok will refactor along these lines, sorry for the misalignment!